### PR TITLE
Added authentication

### PR DESCRIPTION
**Pull request by navvyua**
- User able to login by email and password
- User able to sign up by email, password and password confirmation
- User able to login / sign up via facebook (email takes from it, password generates)
- User able to login / sign up via google (email takes from it, password generates)

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    { uid: auth_params['uid'], provider: auth_params['provider'] }
+  end
+
+  def user_params
+    password = SecureRandom.hex(8)
+    user = User.find_or_create_by(email: auth_params['info']['email']) do |u|
+      u.password = password
+      u.password_confirmation = password
+      # FOR DVELOPMENT saves decrypted password in DB
+      u.password_decrypted = password
```
Comment (IlyaUmanets) [inline app/controllers/sessions_controller.rb]: Why do we need it ?

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  end
+
+  def account_params
+    { uid: auth_params['uid'], provider: auth_params['provider'] }
+  end
+
+  def user_params
+    password = SecureRandom.hex(8)
+    user = User.find_or_create_by(email: auth_params['info']['email']) do |u|
+      u.password = password
```
Comment (IlyaUmanets) [inline app/controllers/sessions_controller.rb]: If user sign in with another provider(with same email) then your script will generate new password for existing user. 
It's wrong way. You should generate password only for new users.
Write specs please.

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  has_secure_password
+
+  has_many :accounts, dependent: :delete_all
+
+  validates_presence_of :email
+  validates :email, uniqueness: true
```
Comment (IlyaUmanets) [inline app/models/user.rb]: Write in one line.
`validates :email, presence: true, uniqueness: true`

File: app/views/sessions/new.html.slim [hunk 4]
```diff
@@ -0,0 +1,14 @@
+h1 Login
+
+= form_tag login_path do
+  p
+    = label_tag :email
+    = email_field_tag :email
```
Comment (IlyaUmanets) [inline app/views/sessions/new.html.slim]: add `required: true`

File: app/views/sessions/new.html.slim [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+  p
+    = label_tag :password
+    = password_field_tag :password
```
Comment (IlyaUmanets) [inline app/views/sessions/new.html.slim]: Same as above.

File: app/views/users/new.html.slim [hunk 6]
```diff
@@ -0,0 +1,11 @@
+h1 Sign Up
+
+= simple_form_for @user do |f|
+  = f.input :email
+  = f.input :password
+  = f.input :password_confirmation
```
Comment (IlyaUmanets) [inline app/views/users/new.html.slim]: Please add `required: true` for all fields.

File: config/application.rb [hunk 7]
```diff
@@ -13,3 +13,5 @@ class Application < Rails::Application
     # -- all .rb files in that directory are automatically loaded.
   end
 end
+
+Rails.application.secrets.each { |key, value| ENV[key.to_s] ||= value }
```
Comment (IlyaUmanets) [inline config/application.rb]: It's wrong way. Use please [figaro](https://github.com/laserlemon/figaro).
You should create config/application.yml and put your keys there.

File: config/routes.rb [hunk 8]
```diff
@@ -1,2 +1,15 @@
 Rails.application.routes.draw do
+  root 'sessions#new'
+
+  get  'signup', to: 'users#new'
+  post 'signup', to: 'users#create', as: 'users'
```
Comment (IlyaUmanets) [inline config/routes.rb]: Do we need this route ?

File: config/routes.rb [hunk 9]
```diff
@@ -1,2 +1,15 @@
...
+
+  get  'login', to: 'sessions#new'
+  post 'login', to: 'sessions#create'
```
Comment (IlyaUmanets) [inline config/routes.rb]: If use signup then write in the same way. Write `signin` or `sign_in` instead of `login`

File: config/routes.rb [hunk 10]
```diff
@@ -1,2 +1,15 @@
...
+
+  delete 'logout', to: 'sessions#destroy'
+
+  get '/auth/:provider/callback', to: 'sessions#create'
+
+  resources :users, except: [:index, :new, :create, :destroy]
```
Comment (IlyaUmanets) [inline config/routes.rb]: In my opinion we should use all user routes.

File: db/migrate/20170114100259_create_users.rb [hunk 11]
```diff
@@ -0,0 +1,14 @@
+class CreateUsers < ActiveRecord::Migration[5.0]
+  def change
+    create_table :users do |t|
+      t.string :name
+      t.string :email
+      t.string :password_digest
+
+      # FOR DEVELOPMENT saves encrypted password in DB
+      t.string :password_decrypted
```
Comment (IlyaUmanets) [inline db/migrate/20170114100259_create_users.rb]: It's unsafe to save passwords without decryption.

File: app/controllers/sessions_controller.rb (hunk 2 above)
Comment (navvyua) [inline app/controllers/sessions_controller.rb]: No, the code block executes only if user was created.
[http://apidock.com/rails/v4.2.7/ActiveRecord/Relation/find_or_create_by](url)

File: app/controllers/sessions_controller.rb (hunk 1 above)
Comment (navvyua) [inline app/controllers/sessions_controller.rb]: Removed.

File: app/models/user.rb (hunk 3 above)
Comment (navvyua) [inline app/models/user.rb]: Done.

File: app/views/sessions/new.html.slim (hunk 4 above)
Comment (navvyua) [inline app/views/sessions/new.html.slim]: Done.

File: app/views/sessions/new.html.slim (hunk 5 above)
Comment (navvyua) [inline app/views/sessions/new.html.slim]: Done.

File: app/views/users/new.html.slim (hunk 6 above)
Comment (navvyua) [inline app/views/users/new.html.slim]: Email is required by default, added for other.

File: config/application.rb (hunk 7 above)
Comment (navvyua) [inline config/application.rb]: Done.

File: config/routes.rb (hunk 8 above)
Comment (navvyua) [inline config/routes.rb]: Sure.

File: config/routes.rb (hunk 9 above)
Comment (navvyua) [inline config/routes.rb]: Edited to sign_up / sign_in / sign_out.

File: config/routes.rb (hunk 10 above)
Comment (navvyua) [inline config/routes.rb]: From my point of view, we don't need index. New and create customized, destroy will be implemented in future.

File: db/migrate/20170114100259_create_users.rb (hunk 11 above)
Comment (navvyua) [inline db/migrate/20170114100259_create_users.rb]: Removed.

**Comment by navvyua** [comment]
I'm writing specs.

File: app/views/users/new.html.slim (hunk 6 above)
Comment (IlyaUmanets) [inline app/views/users/new.html.slim]: @navvyua Ok, Thank you.

File: config/routes.rb (hunk 8 above)
Comment (IlyaUmanets) [inline config/routes.rb]: @navvyua Where do you use this route ?

File: config/routes.rb (hunk 10 above)
Comment (IlyaUmanets) [inline config/routes.rb]: @navvyua Ok, thank you for explanation.

File: config/routes.rb (hunk 8 above)
Comment (navvyua) [inline config/routes.rb]: @IlyaUmanets For registration without fb or google (sign up).

Comment (IlyaUmanets) [inline config/routes.rb]: @navvyua Here app/views/users/new.html.slim ?

Comment (navvyua) [inline config/routes.rb]: @IlyaUmanets Yes.

File: app/views/layouts/application.html.slim [hunk 12]
```diff
@@ -16,3 +16,5 @@ html lang='en'
       = yield
       
     footer
+    
+    / = console
```
Comment (IlyaUmanets) [inline app/views/layouts/application.html.slim]: ?

**Pull request merged**